### PR TITLE
Support array defaults

### PR DESCRIPTION
**Pull request by mikee47**
This PR adds support for `default` values for simple arrrays of Strings or numbers.

A standard object is implemented as a regular C struct, so default initialisation requires a simple `memcpy`.
However, array defaults require some additional code which allocates and loads data.

For now, defaults for Object Arrays are not supported as these are more complex and the benefits less clear.

Example:

```json
"supported_color_models": {
    "type": "array",
    "items": {
        "type":"string"
    },
    "default": ["RGB", "RGBW", "RGBWW", "RAW"]
}
```

Three methods are defined for `Object` (and all derived types):

1. `clear()` resets to default values and any contained arrays are cleared. For a `Union`, this also resets the tag to 0.
2. `loadArrayDefaults()` loads specified defaults for any contained arrays. If no defaults are specified this just clears the arrays.

The `resetToDefaults()` method is provided as a convenient (and slightly more efficient) way to perform (1) and (2) together. This is the action taken if any error occurs (such as 'file not found') when ConfigDB attempts to load a store.

Note that when calling methods such as `importFromFile`, no initialisation is performed beforehand and errors must be dealt with by the application.

### CODE REVIEW

**Comment by mikee47** [comment]
@pljakobs Any thoughts on this PR? I seem to recall you asked a while back about how to reset objects to default. Maybe we add `Object::clear()` to do that?

**Comment by pljakobs** [comment]
I'm a little uncertain about defaults. 
In essence, we have two, maybe three ways to initialize any database:
- defined defaults
- load from file
- programmatic setting
I believe all have their pros. Defined defauts are a fire and forget way to do it. The database would always initialize the way it was defined by the developer.
Load from file is interesting as it is no a whole lot more difficult but adds the ability to have multiple "defaults" - by having multiple files.
Programmatic setting can be required in cases where data can only be generated upon first run.

I guess currently, I'm thinking about using them all but primarily use load from file.
In order to identify, if a database has been initialized at all, I use a property "is_initialized" that defaults to "false", so when I find that value to be false, I know I have to load or set settings. 

So, I don't think that complex defaults are necessary, but at least for simple properties like my "is_initialized" - I would need them.

Resetting is much the same thing, I believe, but for resetting, I would even more tend to use files, as that would give the user the choice of "factory reset" or "last known good configuration" etc.

Having the "factory settings" baked into the database would make that an entirely different process, and being able to reset parts of a database (which is how I would interpret `Object::clear`) might be pretty useful, for example if an array has grown and needs to be pruned back to it's original - on the back end that might also allow to reclaim and defragment storage?

**Comment by mikee47** [comment]
>  if an array has grown and needs to be pruned back to it's original - on the back end that might also allow to reclaim and defragment storage?

Yes actually, arrays are dynamically allocated so doing this *would* reclaim RAM and *probably* defragment, depends on what else used the heap in the meantime...

From what you're saying I think we need both `clear` and `clearToDefaults` method. It could be a `bool` argument to `clear` but I think a separate method is clearer.

`Object::clear()`: reset everything to 0's. This would be the same as if there are no defaults specified in the schema.
`Array::clear()`: empty array, reclaim memory
`Store::clear()`: reset everything to 0's

`Object::clearToDefaults()`: reset everything to configured defaults, including contained arrays.
`Array::clearToDefaults()`: reset array to configured defaults
`Store::clearToDefaults()`: reset everything to configured defaults

So that looks pretty consistent...

Suggestions for method names welcome! (Just not `reset`!)

**Comment by mikee47** [comment]
Having slept on this I'm thinking we should have `clear()` and `resetToDefaults()`.

`Object::clear()`: reset everything to default values, arrays are cleared
`Array::clear()`: empty array, reclaim memory
`Store::clear()`: reset everything to defaults, arrays are cleared

`Object::resetToDefaults()`: reset everything to configured defaults, including contained arrays
`Array::resetToDefaults()`: clear and reset array to configured defaults
`Store::resetToDefaults()`: reset everything to configured defaults, including contained arrays

Note: These will all fail if not called in an update context.

So the difference between the two is purely in how arrays are initialised.

I don't think there is any use case for resetting Object properties to non-default values. It just creates ambiguity and un-neccessary complexity.

To summarise the three initialisation choices you have identified:

1. Defined defaults

When opening a store, `Store::clear()` is first called.
If loading fails because there is no backing file (i.e. not yet initialised), then `Store::resetToDefaults()` is called.

2. load from file

When called from an application this would call `Store::clear()` first.
However, if loading fails then calling `resetToDefaults()` is probably not appropriate - that's up to the application.

3. Programmatic setting

Users can selectively use all available updater methods, including `clear()` and `resetToDefaults()` on stores and objects.

There is also (4) resetting entire database, so we could provide `Database::clear()` and `Database::resetToDefaults()`...

**Comment by mikee47** [comment]
OK, changed my mind again. Let's keep `clear()` methods as above and add `loadArrayDefaults()` as a separate step. I'll be sure to clarify this in the docs.

**Comment by mikee47** [comment]
To recap:

`Array::clear()`: empty array, reclaim memory. For `Object`, `Store` and `ObjectArray`, resets everything to default values, arrays are cleared and memory released.

`Array::loadArrayDefaults()`: Clear and load default entries from schema. For `Object` and `Store` loads defaults for all contained arrays. For `ObjectArray` it's a no-op (for now anyway).

These will all fail if not called in an update context.
